pkg/cmdutil/userio/wizard.go

```diff
@@ -0,0 +1,247 @@
+package userio
+
+import (
+	"fmt"
+
+	"github.com/charmbracelet/bubbles/spinner"
+	tea "github.com/charmbracelet/bubbletea"
+	"github.com/charmbracelet/lipgloss"
+	"github.com/muesli/reflow/wrap"
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+)
+
+type WizardHandler interface {
+	Done()
+	Info(string)
+	SetCurrentTaskCompletedTitle(string)
+	SetInputModel(tea.Model) tea.Model
+	SetTask(string, string)
+	Warn(string)
+}
+
+type asyncWizardHandler struct {
+	messageChan     chan<- tea.Msg
+	inputResultChan <-chan tea.Model
+	done            <-chan bool
+}
+
+func (sh asyncWizardHandler) Done() {
+	sh.update(doneMsg(true))
+	<-sh.done
+}
+
+func (sh asyncWizardHandler) SetInputModel(input tea.Model) tea.Model {
+	sh.update(input)
+	modelResult := <-sh.inputResultChan
+	return modelResult
+}
+
+func (sh asyncWizardHandler) Info(message string) {
+	sh.update(logMsg{
+		level:   zerolog.InfoLevel,
+		message: message,
+	})
+}
+func (sh asyncWizardHandler) Warn(message string) {
+	sh.update(logMsg{
+		level:   zerolog.WarnLevel,
+		message: message,
+	})
+}
+
+func (sh asyncWizardHandler) update(message tea.Msg) {
+	sh.messageChan <- message
+}
+
+func (sh asyncWizardHandler) SetCurrentTaskCompletedTitle(completedTitle string) {
+	sh.update(updateCurrentTaskCompletedTitle(completedTitle))
+}
+
+func (sh asyncWizardHandler) SetTask(title string, completedTitle string) {
+	sh.update(task{title: title, completedTitle: completedTitle, logs: []logMsg{}, completed: false})
+}
+
+func newWizard(streams *IOStreams) WizardHandler {
+	messageChan := make(chan tea.Msg)
+	inputResultChan := make(chan tea.Model)
+	done := make(chan bool)
+
+	m := spinner.New(
+		spinner.WithSpinner(spinner.Dot),
+		spinner.WithStyle(lipgloss.NewStyle().Foreground(lipgloss.Color("#0404ff"))), // CECG Blue
+	)
+	sm := wizardModel{
+		messageChan:     messageChan,
+		inputResultChan: inputResultChan,
+		model:           m,
+		tasks:           []task{},
+	}
+	handler := asyncWizardHandler{
+		messageChan:     messageChan,
+		inputResultChan: inputResultChan,
+		done:            done,
+	}
+
+	go func() {
+		_, err := streams.execute(sm, handler)
+		if err != nil {
+			log.Panic().Msgf("Error in Wizard execution: %s", err.Error())
```

Comment:
Can we log an error or send it to chan to handle by the caller? Having a problem with displaying UI feels like not a reason to panic. I would say: the good reason to panic is either we can't operate at all (which should be decided by top level logic), or something about our internal assertions is wrong (that something is not dependent on external conditions like user or OS)

Reply:
Tiny thing: you can pass error separately like this: `log.Error().Err(err).Msg("Error in Wizard execution")`
And another problem I see here is that it will silently fail in case do not enable logging. But this is minor, I guess